pywr/recorders.py

```diff
@@ -80,59 +80,156 @@ class TablesRecorder(Recorder):
+            an iterable of tuples is provided then the node specific where keyword is
+            used in preference to the where keyword (see below).
+        parameters : iterable or None
+            Parameters to save. Similar to the nodes keyword, except refers to Parameter
+            objects or names thereof.
+        where : string
+            Default path to create the CArrays inside the database.
+        filter_kwds : dict
+            Filter keywords to pass to tables.open_file when opening a file.
+        mode : string
+            Model argument to pass to tables.open_file. Defaults to 'w'
         """
+        filter_kwds = kwargs.pop('filter_kwds', {})
+        mode = kwargs.pop('mode', 'w')
         super(TablesRecorder, self).__init__(model, **kwargs)
+        import tables
 
-        self.parent = parent
+        h5opened = False
+        if isinstance(h5file, basestring):
+            # HACK Python 2 workaround
+            # Note sure how else to deal with str / unicode requirements in pytables
+            # See this issue: https://github.com/PyTables/PyTables/issues/522
+            import sys
+            if sys.version_info[0] == 2 and 'complib' in filter_kwds:
+                filter_kwds['complib'] = filter_kwds['complib'].encode()
+            filters = tables.Filters(**filter_kwds)
+            h5fh = tables.open_file(h5file, mode=mode, filters=filters)
+            h5opened = True
+        elif isinstance(h5file, tables.File):
+            h5fh = h5file
+        else:
+            raise ValueError("Argument h5file must be either a filename string or instance of tables.File")
+
+        self.h5file = h5fh
+        self._h5opened = h5opened
         self.nodes = nodes
+        self.parameters = parameters
+        self.where = where
 
         self._arrays = None
 
     def __del__(self):
+        if self._h5opened and self.h5file.isopen:
+            self.h5file.close()
         if self._arrays is not None:
             for arr in self._arrays.values():
                 arr.close()
-        del(self._arrays)
+            del self._arrays
+
+    def finish(self):
+        if self._h5opened:
+            self.h5file.close()
 
+    @classmethod
+    def load(cls, model, data):
+        import os
+        url = data.pop("url")
+        if not os.path.isabs(url) and model.path is not None:
+            url = os.path.join(model.path, url)
+        return cls(model, url, **data)
 
     def setup(self):
         """
         Setup the tables
         """
+        from pywr.parameters import IndexParameter
         import tables
         shape = len(self.model.timestepper), len(self.model.scenarios.combinations)
-        atom = tables.Float64Atom()
         # Create a CArray for each node
         self._arrays = {}
 
         # Default to all nodes if None given.
         if self.nodes is None:
-            nodes = self.model.node.values()
+            nodes = [(self.where, n) for n in self.model.node.values()]
```

Comment:
Please use `self.model.nodes` instead of `self.model.node`. (check for other instances)